File: backend-node/src/parsers/file_parser.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

class EmployeeParser {
  constructor() {
    this.REQUIRED_COLS = new Set(['Employee_Name', 'Employee_EmailID']);
  }

  parseFile(filepath) {
    if (!fs.existsSync(filepath)) {
      throw new Error(`File not found: ${filepath}`);
    }

    const ext = path.extname(filepath).toLowerCase();
    if (ext === '.xlsx' || ext === '.xlsm' || ext === '.xls') {
      return this._parseExcel(filepath);
    }
    if (ext === '.csv' || ext === '.tsv') {
      return this._parseCsv(filepath);
    }
    throw new Error(`Unsupported file format: ${ext}`);
  }

  _parseCsv(filepath) {
    const workbook = XLSX.readFile(filepath);
    const sheetName = workbook.SheetNames[0];
    const df = XLSX.utils.sheet_to_json(workbook.Sheets[sheetName]);
    return this._toEmployees(df);
  }

  _parseExcel(filepath) {
    const workbook = XLSX.readFile(filepath);
    const sheetName = workbook.SheetNames[0];
    const df = XLSX.utils.sheet_to_json(workbook.Sheets[sheetName]);
    return this._toEmployees(df);
  }

  _toEmployees(df) {
    const columns = new Set(Object.keys(df[0] || {}));
    const missing = [...this.REQUIRED_COLS].filter(c => !columns.has(c));
    if (missing.length > 0) {
      throw new Error(`Missing required columns: ${missing.join(', ')}`);
    }

    const { Employee } = require('../models/employee');
    return df
      .filter(row => row.Employee_Name && row.Employee_EmailID)
      .map(row => new Employee({
        name: String(row.Employee_Name),
        email: String(row.Employee_EmailID)
      }));
  }
}

class AssignmentParser {
  constructor() {
    this.REQUIRED_COLS = new Set([
      'Employee_Name', 'Employee_EmailID',
      'Secret_Child_Name', 'Secret_Child_EmailID'
    ]);
  }

  parseFile(filepath) {
    if (!fs.existsSync(filepath)) {
      throw new Error(`File not found: ${filepath}`);
    }

    const ext = path.extname(filepath).toLowerCase();
    let df;
    if (ext === '.xlsx' || ext === '.xlsm' || ext === '.xls') {
      const workbook = XLSX.readFile(filepath);
      const sheetName = workbook.SheetNames[0];
      df = XLSX.utils.sheet_to_json(workbook.Sheets[sheetName]);
    } else {
      const workbook = XLSX.readFile(filepath);
      const sheetName = workbook.SheetNames[0];
      df = XLSX.utils.sheet_to_json(workbook.Sheets[sheetName]);
    }

    const columns = new Set(Object.keys(df[0] || {}));
    const missing = [...this.REQUIRED_COLS].filter(c => !columns.has(c));
    if (missing.length > 0) {
      throw new Error(`Missing required columns: ${missing.join(', ')}`);
    }

    const { Assignment, Employee } = require('../models/employee');
    return df
      .filter(row => row.Employee_Name && row.Employee_EmailID &&
                     row.Secret_Child_Name && row.Secret_Child_EmailID)
      .map(row => new Assignment({
        giver: new Employee({
          name: String(row.Employee_Name),
          email: String(row.Employee_EmailID)
        }),
        receiver: new Employee({
          name: String(row.Secret_Child_Name),
          email: String(row.Secret_Child_EmailID)
        })
      }));
  }
}

module.exports = { EmployeeParser, AssignmentParser };

File: backend-python/src/parsers/file_parser.py
"""
Parsers for reading employee and previous-assignment data from CSV / XLSX files.
"""
import csv
import io
from pathlib import Path
from typing import List, Tuple

import pandas as pd

from src.models.employee import Employee, Assignment


class EmployeeParser:
    """Parses employee lists from CSV or XLSX files."""

    REQUIRED_COLS = {"Employee_Name", "Employee_EmailID"}

    def parse_file(self, filepath: str) -> List[Employee]:
        """Auto-detect format and parse the employee list."""
        path = Path(filepath)
        if not path.exists():
            raise FileNotFoundError(f"File not found: {filepath}")
        if path.suffix.lower() in (".xlsx", ".xlsm", ".xls"):
            return self._parse_excel(path)
        if path.suffix.lower() in (".csv", ".tsv"):
            return self._parse_csv(path)
        raise ValueError(f"Unsupported file format: {path.suffix}")

    def parse_csv_content(self, content: str) -> List[Employee]:
        """Parse employee list from a raw CSV string (e.g. from API upload)."""
        reader = csv.DictReader(io.StringIO(content))
        self._validate_columns(set(reader.fieldnames or []))
        return [
            Employee(name=row["Employee_Name"], email=row["Employee_EmailID"])
            for row in reader
            if row.get("Employee_Name") and row.get("Employee_EmailID")
        ]

    def _parse_csv(self, path: Path) -> List[Employee]:
        df = pd.read_csv(path)
        self._validate_columns(set(df.columns))
        return self._df_to_employees(df)

    def _parse_excel(self, path: Path) -> List[Employee]:
        df = pd.read_excel(path, engine="openpyxl")
        self._validate_columns(set(df.columns))
        return self._df_to_employees(df)

    @staticmethod
    def _df_to_employees(df: pd.DataFrame) -> List[Employee]:
        df = df.dropna(subset=["Employee_Name", "Employee_EmailID"])
        return [
            Employee(name=str(row["Employee_Name"]), email=str(row["Employee_EmailID"]))
            for _, row in df.iterrows()
        ]

    def _validate_columns(self, columns: set):
        missing = self.REQUIRED_COLS - columns
        if missing:
            raise ValueError(f"Missing required columns: {missing}")


class AssignmentParser:
    """Parses previous-year Secret Santa assignments from CSV / XLSX files."""

    REQUIRED_COLS = {
        "Employee_Name", "Employee_EmailID",
        "Secret_Child_Name", "Secret_Child_EmailID",
    }

    def parse_file(self, filepath: str) -> List[Assignment]:
        path = Path(filepath)
        if not path.exists():
            raise FileNotFoundError(f"File not found: {filepath}")
        if path.suffix.lower() in (".xlsx", ".xlsm", ".xls"):
            df = pd.read_excel(path, engine="openpyxl")
        else:
            df = pd.read_csv(path)
        self._validate_columns(set(df.columns))
        return self._df_to_assignments(df)

    def parse_csv_content(self, content: str) -> List[Assignment]:
        reader = csv.DictReader(io.StringIO(content))
        self._validate_columns(set(reader.fieldnames or []))
        return [self._row_to_assignment(row) for row in reader]

    @staticmethod
    def _df_to_assignments(df: pd.DataFrame) -> List[Assignment]:
        df = df.dropna(subset=["Employee_Name", "Employee_EmailID",
                                "Secret_Child_Name", "Secret_Child_EmailID"])
        return [
            Assignment(
                giver=Employee(name=str(r["Employee_Name"]), email=str(r["Employee_EmailID"])),
                receiver=Employee(name=str(r["Secret_Child_Name"]), email=str(r["Secret_Child_EmailID"])),
            )
            for _, r in df.iterrows()
        ]

    @staticmethod
    def _row_to_assignment(row: dict) -> Assignment:
        return Assignment(
            giver=Employee(name=row["Employee_Name"], email=row["Employee_EmailID"]),
            receiver=Employee(name=row["Secret_Child_Name"], email=row["Secret_Child_EmailID"]),
        )

    def _validate_columns(self, columns: set):
        missing = self.REQUIRED_COLS - columns
        if missing:
            raise ValueError(f"Missing required columns: {missing}")
